db/index.js

```diff
@@ -0,0 +1,79 @@
+const firebase = require('firebase');
+const config = require('../config.js');
+
+if (!firebase.apps.length) {
+	firebase.initializeApp(config.config);
+}
+
+const db = firebase.database().ref("live_weller").child("markets");
+
+/**
+ * This method returns a reference to the most recent version of the database.
+ * Using the result of this method and .val(), you can access the entire
+ * list of markets in JSON format.
+ */
+function getAllMarkets() {
```

Comment:
I tried doing this earlier and was having trouble.  The reason it's set up like this is because its asynchronous, so I was having difficulty returning the value when it finally arrived.  To get around this I had been doing .once('value').then(function(...){}), although this only worked since I was doing everything inside the .then function to catch the data.  

Here I was working with trying to build the JSON myself since the .then no longer worked like this, but I could not get it to return pure JSON.  If I put a .then inside that returns the results.val() value, then it still returns a Snapshot object that I have to use .val() with to extract the information.

Do you have any tips on what to do here? I tried working with Promises but they weren't working out too hot and I was just having trouble with them in general.

Reply:
The problem might be because it's an asynchronous function that you're calling in synchronous code.
```javascript
// block 1
function getAllMarkets() {
    return db.once('value').then(result => result.val());
}

// block 2
const markets = getAllMarkets();
console.log(markets);
```

Block 1 has an asynchronous operation and a success callback, but block 2 is still synchronous code, so the value of the `getAllMarkets` call might be getting assigned to `const markets` before the callback finishes. Try returning your own promise, something like
```javascript
function getAllMarkets() {
    return new Promise((resolve, reject) => {
        db.once('value').then(result => resolve(result.val())).catch(err => reject(err));
    });
}

getAllMarkets().then(markets => {
    console.log(markets);
});
```

Reply:
That works so great, thanks so much.  This should hopefully be wrapped up tonight.

Reply:
Nice! Do you understand how it works/why it didn’t work earlier? Async can be kinda tricky at first but it’s super foundational.

Reply:
I think I get what I did wrong, I think I was making a promise in the route that was based on waiting for getAllMarkets.  Was just mixed up on the declaration and returning, so this really cleared it up.